oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +239,282 @@ def all(self, length, keyword):
+    def expire_all(self):
+        for response_type in self._cache_lifecycle.keys():
+            for key_hash in self._cache_lifecycle[response_type].keys():
+                yield self.expire(response_type, key_hash)
+
+    @defer.inlineCallbacks
+    def cache_value(self, response_type, key, value):
+        if response_type not in self._response_types:
+            raise Exception("Invalid response type")
+        if self.enable_caching:
+            key_hash = sha256(key).hexdigest()
+            cache_file = os.path.join(self._cache_dir, response_type, key_hash)
+
+            if key_hash in self._cache_lifecycle[response_type]:
+                yield self.expire(response_type, key, value)
+
+            self._cache_lifecycle[response_type][key_hash] = {
+                'expiration': reactor.callLater(self.expiration_time,
+                                                self.expire,
+                                                response_type, key_hash),
+                'lock': defer.DeferredLock()
+            }
+            lock = self._cache_lifecycle[response_type][key_hash]['lock']
+            yield lock.acquire()
+            with open(cache_file, 'w+') as fw:
+                json.dump(value, fw)
+            lock.release()
+
+    @defer.inlineCallbacks
+    def expire(self, response_type, key_hash):
+        if response_type not in self._response_types:
+            raise Exception("Invalid response type")
+        lifecycle = self._cache_lifecycle[response_type][key_hash]
+        if lifecycle['expiration'].active():
+            lifecycle['expiration'].cancel()
+
+        yield lifecycle['lock'].acquire()
+        try:
+            os.remove(os.path.join(self._cache_dir, response_type, key_hash))
+        except OSError:
+            pass
+        lifecycle['lock'].release()
+        del self._cache_lifecycle[response_type][key_hash]
+
+    @defer.inlineCallbacks
+    def lookup(self, response_type, key):
+        if not self.enable_caching:
+            defer.returnValue(None)
+
+        key_hash = sha256(key).hexdigest()
+        cache_file = os.path.join(self._cache_dir, response_type, key_hash)
+
+        if key_hash not in self._cache_lifecycle[response_type]:
+            defer.returnValue(None)
+
+        lock = self._cache_lifecycle[response_type][key_hash]['lock']
+        expiration = \
+            self._cache_lifecycle[response_type][key_hash]['expiration']
+
+        yield lock.acquire()
+
+        if not os.path.exists(cache_file):
+            lock.release()
+            defer.returnValue(None)
+
+        with open(cache_file, 'r') as fh:
+            value = json.load(fh)
+
+        expiration.reset(self.expiration_time)
+        lock.release()
+        defer.returnValue(value)
+
+    @defer.inlineCallbacks
+    def http_request(self, url):
+        cached_value = yield self.lookup('http_request', url)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
```

Comment:
Yeah, IRC what it does is it raises a `StopIteration` exception so that the generator is interrupted and the value passed to return value is passed as a callback argument to the deferred.